### PR TITLE
Adding a colorizer to bot when loading new video.

### PR DESCRIPTION
**Pull request by atrivedi**
The color is a darker version of the user that prompted the change

#49

### CODE REVIEW

File: lib/bot.js [hunk 1]
```diff
@@ -56,7 +56,9 @@ function Bot(config, gapi) {
     gapi.video_specs(video_obj, function(obj) {
       if (obj.title) {
         var msg = obj.user + ' started "' + obj.title + '"'
-        callback(bot_msg_obj(msg));
+        var msg_obj = bot_msg_obj(msg);
+        msg_obj.user = obj.user;
```
Comment (aromatt) [inline lib/bot.js]: part of `bot_msg_obj` is setting the user to the bot -- we want the end user to see these messages come from him.

File: lib/chat.js [hunk 2]
```diff
@@ -100,8 +100,15 @@ function Chat(config, bot) {
 
   // When there's a new video state, tell the bot and broadcast what he says
   this.video_started = function(new_state) {
-    bot.video_started(new_state, this.broadcast_chat);
-  }
+    var startedCallback = function(msg_obj) {
+      if(msg_obj.hasOwnProperty('user')) {
+        msg_obj.customPanelColor = this.color_for(msg_obj.user);
```
Comment (aromatt) [inline lib/chat.js]: so we'll need to keep `new_state.user` and `msg_obj.user` distinct, and use `new_state.user` for the custom panel color.

File: views/chat.jade [hunk 3]
```diff
@@ -141,23 +141,42 @@ script.
...
+              msg_obj.customPanelColor;
+      var background = chat_li.css('background');
+      var rgb = color.hexToRgb(alias);
+      try {
+        var hsl = color.rgbToHsl( rgb.r, rgb.g, rgb.b );
+        hsl[2] = hsl[2] * 0.25; //quarter of the brightness
+        var changedRgb = color.hslToRgb.apply( null, hsl );
+        background = color.rgbToHex.apply( null, changedRgb );
+        console.log( 'oldColor: ' + msg_obj.customPanelColor + '\nnew color: ' + background );
+      } catch(err) {
```
Comment (aromatt) [inline views/chat.jade]: I appreciate the precaution

**Comment by aromatt** [comment]
I like the overall change though, looks tasty

**Comment by aromatt** [comment]
if anything I think we can make it even more subdued, like just *barely* tilt the usual grey toward the user's color

**Comment by aromatt** [comment]
Another option to make it more subtle but closer to the user's actual color would be to use a border instead of coloring the whole panel for the message

**Comment by atrivedi** [comment]
@aromatt I've updated it with the suggestions. tell me what you think. We could certainly add a border

File: views/chat.jade [hunk 4]
```diff
@@ -147,24 +147,44 @@ script.
...
       this.last_emote = msg_obj.emote;
       this.last_user = msg_obj.user;
-      var chat_li = new_chat_li(msg_obj);
+      chat_li = new_chat_li(msg_obj);
       list.append(chat_li);
     } else {
+      chat_li = list.find('li').last()
       for (line of msg_obj.text.split('\n')) {
-        list.find('li').last().append('<br>');
+        chat_li.append('<br>');
```
Comment (aromatt) [inline views/chat.jade]: Why these changes?  It's supposed to append to the same panel if it's the same user

File: views/chat.jade [hunk 5]
```diff
@@ -147,24 +147,44 @@ script.
...
-          list.find( 'li' ).last().append( makeLink(line) );
+          chat_li.append( makeLink( line ) );
         } else {
-          list.find( 'li' ).last().append( document.createTextNode( line ) );
+          chat_li.append( document.createTextNode( line ) );
         }
       }
     }
-
+    if(msg_obj.hasOwnProperty('customPanelColor')) {
```
Comment (aromatt) [inline views/chat.jade]: This seems like it could be in `new_chat_li()`

File: views/chat.jade (hunk 4 above)
Comment (atrivedi) [inline views/chat.jade]: it will append do the same panel. `chat_li = list.find('li').last()` is the last panel. I did this way so that, in the future, if we ever want to custom color a panel again, we just have to add that property. Here, we can color on triggers that aren't the first new_li

File: views/chat.jade (hunk 5 above)
Comment (atrivedi) [inline views/chat.jade]: mentioned above: it's here so that we can use this in the future if needed. that way we aren't forced to just change color on the first `new_li`

**Comment by atrivedi** [comment]
now that im thinknig about it. the "desaturation" logic should be in the `.js` file . the `.jade` file should strictly set the property. Thoughts?

File: views/chat.jade (hunk 4 above)
Comment (aromatt) [inline views/chat.jade]: again I'm sorry I'm not paying much attention, I'll have to come back to this later

Comment (aromatt) [inline views/chat.jade]: ok, yeah I misread your change - this is better

File: views/chat.jade (hunk 5 above)
Comment (aromatt) [inline views/chat.jade]: From what I can tell, putting this here instead of in `new_chat_li()` opens the door for changing the color of a chat li that has already been drawn -- is that what you're saying?  From a UX perspective I'm not sure I want to allow that, which is why I think it should go into that function... If I'm missing the point here, let me know.

Comment (atrivedi) [inline views/chat.jade]: You are correct. What if the bot notifies two song changes back to back? Would you expect the color of the original or the color of the latest. Furthermore, should we not color the `<li>`, but rather the children of the `<li>` to account for this? idk

Comment (aromatt) [inline views/chat.jade]: Well, currently if the bot sends any two messages back to back, it's two separate `<li>`s (see https://github.com/aromatt/thejub.pub/blob/master/views/chat.jade#L150).  This makes it easier to pick out jub each message and is important for song-start notifications so that we see a timestamp for each song.

So in the new, colored world, we'd get a separate `<li>` for each song notification, each with the respective DJ's color.  Thoughts?

**Comment by aromatt** [comment]
re: moving desaturation logic, yeah that makes sense to me -- anything that might be useful to other sections of the UI could go into `lib.js`

File: views/chat.jade (hunk 5 above)
Comment (atrivedi) [inline views/chat.jade]: yeah, that makes sense. I'll move it the new_li thing

**Comment by atrivedi** [comment]
@aromatt updated!

File: lib/chat.js [hunk 6]
```diff
@@ -96,12 +97,30 @@ function Chat(config, bot) {
...
-  }
+  };
 
   // When there's a new video state, tell the bot and broadcast what he says
   this.video_started = function(new_state) {
-    bot.video_started(new_state, this.broadcast_chat);
-  }
+    var startedCallback = function(msg_obj) {
+      var userColor = this.color_for(new_state.user);
+      var alias = color.colorNameToHex(userColor) ?
```
Comment (aromatt) [inline lib/chat.js]: out of curiosity... would this do the same thing?
`var alias = color.colorNameToHex(userColor) || userColor`

**Comment by aromatt** [comment]
this works like I think it should and the colors look good.  Let's try this out and see what people think!

File: lib/chat.js (hunk 6 above)
Comment (atrivedi) [inline lib/chat.js]: yup! The next time i hit this file, i'll change it
